File: labs/mobile/donor-archive/tunnel/PublicRelayAggregator.kt
package com.luminet.android.tunnel

data class AndroidRelayNode(
    val nodeId: String,
    val protocol: String,
    val host: String,
    val port: Int,
    val countryCode: String,
    var pingMs: Int,
    var isAlive: Boolean = true
)

class PublicRelayAggregator {
    private val relays = mutableMapOf<String, AndroidRelayNode>()

    fun ingestNode(node: AndroidRelayNode) {
        relays[node.nodeId] = node
    }

    fun updateHealth(nodeId: String, pingMs: Int, isAlive: Boolean) {
        val n = relays[nodeId] ?: return
        n.pingMs = pingMs
        n.isAlive = isAlive
    }

    fun queryRelays(country: String?, proto: String?, maxResults: Int): List<AndroidRelayNode> {
        return relays.values
            .filter { it.isAlive }
            .filter { country == null || it.countryCode == country }
            .filter { proto == null || it.protocol == proto }
            .sortedBy { it.pingMs }
            .take(maxResults)
    }

    fun totalCount(): Int = relays.size
}

File: labs/control-ui/api-shadow/public_relay_aggregator.ts
export interface UiRelayNode {
  nodeId: string;
  protocol: string;
  host: string;
  port: number;
  countryCode: string;
  pingMs: number;
  isAlive?: boolean;
}

export class PublicRelayAggregator {
  private relays: Map<string, UiRelayNode> = new Map();

  ingestNode(node: UiRelayNode): void {
    this.relays.set(node.nodeId, { ...node, isAlive: node.isAlive ?? true });
  }

  updateHealth(nodeId: string, pingMs: number, isAlive: boolean): boolean {
    const n = this.relays.get(nodeId);
    if (!n) return false;
    n.pingMs = pingMs;
    n.isAlive = isAlive;
    return true;
  }

  queryRelays(country?: string, proto?: string, maxResults: number = 10): UiRelayNode[] {
    const list = Array.from(this.relays.values())
      .filter(n => n.isAlive)
      .filter(n => !country || n.countryCode === country)
      .filter(n => !proto || n.protocol === proto)
      .sort((a, b) => a.pingMs - b.pingMs);

    return list.slice(0, maxResults);
  }

  totalCount(): number {
    return this.relays.size;
  }
}
